### PR TITLE
Issue-115 Made the magazine badge dynamic

### PR DESCRIPTION
**Pull request by aabi01**
Issue#115 is resolved. I have made the magazine badge dynamic using Laravel Cache.

### CODE REVIEW

File: app/Http/Controllers/PhpReboot/HomeController.php [hunk 1]
```diff
@@ -14,6 +17,29 @@ class HomeController extends Controller
...
+        //Number of Articles
+        $articlesCount = Cache::remember('articlesCount', $cacheTimeout, function(){
+            return Article::count();
+        });
+        
+        //Last Article Date
+        $lastArticleDate = Cache::remember('lastArticleDate', $cacheTimeout, function(){
+            $lastArticle = Article::orderBy('published_at', 'Desc')->first();
+            return $lastArticle->published_at;
+        });
```
Comment (kapilsharma) [inline app/Http/Controllers/PhpReboot/HomeController.php]: Hi @aabi01 ,

Thanks for pull request. Please do not mind but this is not the right way of doing it. Let me explain:

**Thin controller**

Laravel follows MVC pattern. In MVC, we says we must have thin controller & fat model. Please note, here models are not exactly eloquent model but a model layer, which contains services, Repository & Model (Eloquent classes)

**Benefit of thin controller**

MVC or all other patterns basically says about **separation of concerns**. It exactly means that one class (or method) should be responsible for only one functionality. Do only one thing but do it completely and in best way. Same laws are given by **SOLID** principles where

- S - Single responsibility principle. Here our controller have multiple responsibilities (actually all)
- O - Open Close Principle, means our code is open for extension but close for modification. Here, if we need to change file cache to say memcache in future, we will need to update our controller. Thus my code is not closed for modification.
- L - Leskov Substitution Principle, means program should be replaceable with instances of their subtypes. Here, Cache is interface and cache drivers (file cache, memcache, Redis etc) are its instance (Concrete class). Here, we can not change concrete class without changing code and thus we are not following this principle.
- I - Interface segregation principle, not applicable here.
- D - Dependency Inversion Principle. Laravel provide it through IOC container (Service containers/Providers), which we are not using. Thus not following it.

Don't worry, it is still a very good attempt. I can't accept this PR but I'll take code from your repo and write it following all above principles. You can then relate my solution and your and confirm how we can do it in better way. I'm going to Diwali holidays from tomorrow so might need some time to implement it but I'll try to do it ASAP.

**Few links for more study**: (You can fine more using Google)

Thin controller: https://adamwathan.me/2013/10/02/putting-your-laravel-controllers-on-a-diet/
SOLID - https://en.wikipedia.org/wiki/SOLID_(object-oriented_design)

**Comment by kapilsharma** [comment]
Please do not make changes your issue-115 branch, I'll take code from there and try to keep it actual to maximum extent.

You can do one thing, make same PR to new `issue-115` branch. I'll merge it there and work on that branch.

File: app/Http/Controllers/PhpReboot/HomeController.php (hunk 1 above)
Comment (aabi01) [inline app/Http/Controllers/PhpReboot/HomeController.php]: Hi @kapilsharma ,
No problem with that. I am glad you pointed out the mistakes. Thank you very much for the review. I will surely take them into consideration next time. I will be waiting to see your solution.
Now, i will make a new PR to issue-115 branch.

Happy Diwali !!!

**Comment by aabi01** [comment]
Hi @kapilsharma ,
I am trying to make PR to `issue-115` branch on upstream but i can't find a branch with that name there.

**Comment by kapilsharma** [comment]
Ahh, may be you didn't have permission for creating new branch. I created `issue-115` branch from development. Please make PR now.

**Comment by aabi01** [comment]
Hi @kapilsharma what about this issue? havn't you found the time yet?
you can let me know which files to change, i will be happy to implemment this change.

**Comment by kapilsharma** [comment]
Hi @aabi01 

Sorry for delay but yes, During week long Diwali holiday, I was little too busy with-in family and only have limited internet connectivity through my mobile. Then first two days in office after a week-off were too hectic.

However now I put technical approach in the issue. I guess fixing issue should have been easier then writing big comment there but its your issue so I wish you to understand the reasons and commit it. I just provided my technical approach. You are welcome to provide any suggestion or ask any doubt.

**Comment by kapilsharma** [comment]
@aabi01 I'm closing this PR. Please create new PR based on tech approach we discussed in issue. Obviously, if you have any doubts or suggestion in my tech approach, we can discuss further.

**Pull request closed without merging**